### PR TITLE
Gltf material properties are now passed to the runtime

### PR DESCRIPTION
**Pull request by JCash**
The base properties of a gltf materials are now stored with the runtime model.
This allows us to better render the model as it was authored.

We support the material properties/extensions listed here:
https://github.com/jkuhlmann/cgltf?tab=readme-ov-file#features

# Technical details

Once the engine/tests are built, you can test a model like so:

After building `modelc`:

C++ (good for testing/debugging with ASan etc)
```
engine/modelc/build/src/test/test_model DamagedHelmet.gltf
```

Java/Jni
```
engine/modelc/scripts/test_model_importer.sh DamagedHelmet.gltf
```

After `build_bob_light`:
```
engine/modelc/scripts/test_model_importer.sh DamagedHelmet.gltf
```

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/modelc/src/modelimporter_jni.cpp [hunk 1]
```diff
@@ -65,14 +148,14 @@ static jobjectArray CreateMaterialsArray(JNIEnv* env, dmModelImporter::jni::Type
     for (uint32_t i = 0; i < count; ++i)
     {
         const dmModelImporter::Material* material = &materials[i];
-        jobject obj = CreateMaterial(env, types, material);
+        jobject obj = C2J_CreateMaterial(env, types, material);
```
Comment (JCash) [inline engine/modelc/src/modelimporter_jni.cpp:151]: Here I chose to _not_ use the object caching mechanism, as these structs are so many to manually maintain.
And they should also be small enough, and not as reused so it shouldn't be a huge deal.

The caveat with this type function is, until I've fixed the object cache feature, is that you cannot do it with cyclic dependencies, as it doesn't know when to sto creating new objects.
Once fixed, we could potentially create the entire scene with just one generated call.

File: engine/rig/proto/rig/rig_ddf.proto [hunk 2]
```diff
@@ -8,6 +8,174 @@ option java_package = "com.dynamo.rig.proto";
...
 
+enum AlphaMode {
+    OPAQUE  = 0;
+    MASK    = 1;
+    BLEND   = 2;
+    MAX_ENUM= 3;
+}
+
+// We're not really using the Image concept at runtime
+// message Image
```
Comment (JCash) [inline engine/rig/proto/rig/rig_ddf.proto:19]: left for clarification

File: engine/rig/proto/rig/rig_ddf.proto [hunk 3]
```diff
@@ -8,6 +8,174 @@ option java_package = "com.dynamo.rig.proto";
...
+
+message Texture
+{
+    optional string name    = 1;
+    optional uint32 index   = 2; // Index into the scene textures
+    optional string path    = 3; // Technically an Image struct
+    optional Sampler sampler = 4;
+}
+
+// KHR_texture_transform: https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_texture_transform/README.md
```
Comment (JCash) [inline engine/rig/proto/rig/rig_ddf.proto:47]: Tried to be diligent of adding the specifications.

File: scripts/unpack_ddf.py [hunk 4]
```diff
@@ -162,6 +162,10 @@ def print_object(printer, msg):
     for descriptor in msg.DESCRIPTOR.fields:
         value = getattr(msg, descriptor.name)
 
+        if not descriptor.label == descriptor.LABEL_REPEATED:
+            if not msg.HasField(descriptor.name):
+                continue
```
Comment (JCash) [inline scripts/unpack_ddf.py:167]: At runtime, we'll need to check if the structure is actually present in the data, using the "hasX" functionality.

**Review by Jhonnyg — DISMISSED** [review]
Excellent! Onward and upwards!

File: engine/rig/proto/rig/rig_ddf.proto [hunk 5]
```diff
@@ -8,6 +8,174 @@ option java_package = "com.dynamo.rig.proto";
 option java_outer_classname = "Rig";
 
 
+enum AlphaMode {
+    OPAQUE  = 0;
+    MASK    = 1;
+    BLEND   = 2;
+    MAX_ENUM= 3;
+}
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/rig/proto/rig/rig_ddf.proto:16]: Yes:
https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#_material_alphamode

File: engine/rig/proto/rig/rig_ddf.proto [hunk 6]
```diff
@@ -8,6 +8,174 @@ option java_package = "com.dynamo.rig.proto";
...
+// }
+
+message Sampler
+{
+    optional string name        = 1;
+    optional uint32 index       = 2;                      // Index into the scene samplers list
+    optional uint32 magFilter   = 3 [default = 9728]; // Required=No, No default, NEAREST=9728, LINEAR=9729
+    optional uint32 minFilter   = 4 [default = 9728]; // Required=No, No default, NEAREST=9728, LINEAR=9729
+    optional uint32 wrapS       = 5 [default = 10497];    // Required=No, Default=10497 (REPEAT)
+    optional uint32 wrapT       = 6 [default = 10497];    // Required=No, Default=10497 (REPEAT)
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/rig/proto/rig/rig_ddf.proto:36]: We could probably do that.
I couldn't really find an enum somewhere but "names" such as "NEAREST".
https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#_sampler_minfilter
It's a bit strange tbh.

Let's do that for the next iteration then.

**Pull request merged**